sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datagen/DataGeneratorRowFn.java

```diff
@@ -0,0 +1,189 @@
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Random;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.extensions.sql.impl.utils.CalciteUtils;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.beam.vendor.calcite.v1_28_0.org.apache.commons.lang.RandomStringUtils;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** A stateful DoFn that converts a sequence of Longs into structured Rows. */
+public class DataGeneratorRowFn extends DoFn<Long, Row> {
+  private final Schema schema;
+  private final ObjectNode properties;
+  private final @Nullable String primaryTimestampField;
+
+  private transient Map<String, FieldGenerator> fieldGenerators;
+  private transient Random random;
+
+  @SuppressWarnings("initialization")
+  public DataGeneratorRowFn(
+      Schema schema, ObjectNode properties, @Nullable String primaryTimestampField) {
+    this.schema = schema;
+    this.properties = properties;
+    this.primaryTimestampField = primaryTimestampField;
+  }
+
+  @Setup
+  public void setup() {
+    this.random = new Random();
+    this.fieldGenerators = new HashMap<>();
+
+    for (Schema.Field field : schema.getFields()) {
+      fieldGenerators.put(field.getName(), createGeneratorForField(field));
+    }
+  }
+
+  @ProcessElement
+  public void processElement(
+      @Element Long index, @Timestamp Instant timestamp, OutputReceiver<Row> out) {
+    Row.Builder rowBuilder = Row.withSchema(schema);
+    for (Schema.Field field : schema.getFields()) {
+      Object value;
+      if (field.getName().equals(this.primaryTimestampField)) {
+        value = timestamp.toDateTime();
+      } else {
+        FieldGenerator generator = fieldGenerators.get(field.getName());
+        if (generator == null) {
+          throw new IllegalStateException("No generator found for field: " + field.getName());
+        }
+        value = generator.generate(index);
+      }
+      rowBuilder.addValue(value);
+    }
+    out.output(rowBuilder.build());
+  }
+
+  @FunctionalInterface
+  private interface FieldGenerator extends Serializable {
+    @Nullable
+    Object generate(long index);
+  }
+
+  private FieldGenerator createGeneratorForField(Schema.Field field) {
+    String fieldName = field.getName();
+    FieldGenerator valueGenerator = createValueGeneratorForField(field);
+    double nullRate = properties.path("fields." + fieldName + ".null-rate").asDouble(0.0);
+
+    if (nullRate > 0) {
+      return (index) ->
+          Objects.requireNonNull(random).nextDouble() < nullRate
```

Comment:
Consider using `ThreadLocalRandom.current()` instead of `new Random()`.

Reply:
Same for other uses of `random` in the `createValueGeneratorForField()` method below